### ISSUE TITLE
customImportOrder does not work

### ISSUE THREAD

**Issue by appraveen** [jshiell/checkstyle-idea#377]
Thank you @mqzry for the feature.  There is a bug from the pull #363 which prevents customImportOrder to work properly.

```
private void parseImportOrderRules(@NotNull String value) {
    String[] groups = value.split(value);
    customImportOrder = Arrays.stream(groups).map(ImportGroup::valueOf).collect(Collectors.toList());
}
```

Using `value.split("###")` might fix the issue. [Spec].(http://checkstyle.sourceforge.net/apidocs/com/puppycrawl/tools/checkstyle/checks/imports/CustomImportOrderCheck.html)

**Comment by mqzry** [comment]
Thanks for reporting. I'll fix it and create some tests the coming days.

**Comment by appraveen** [comment]
Looks like #372 change is already released in **5.16.3**. But, the below configuration is not working in Intellij IDEA 2017.3.3

```
 <module name="CustomImportOrder">
            <property name="standardPackageRegExp" value="^(java|javax)\."/>
            <property name="specialImportsRegExp" value="^org\."/>
            <property name="sortImportsInGroupAlphabetically" value="true"/>
            <property name="separateLineBetweenGroups" value="true"/>
            <property name="customImportOrderRules" value="STANDARD_JAVA_PACKAGE###SPECIAL_IMPORTS###THIRD_PARTY_PACKAGE###STATIC"/>
        </module>
```

**Comment by yeoji** [comment]
@appraveen Is there an error? What doesn't work?

I just tested using the configuration you have provided and it seems alright. Imported correctly and I had no errors. I am on IntelliJ 2017.2.5 though, not sure if that matters.

**Comment by appraveen** [comment]
I am not getting any error, when I import, it says, "Checkstyle configuration settings were imported to Default copy". I don't see the order reflected in the dialog and reformatting the code does not pick up the order mentioned in checkstyle.xml

Checstyle -IDEA 
![image](https://user-images.githubusercontent.com/2443047/36008596-efb919d6-0cfc-11e8-9f26-c4608d709506.png)


Code Style Preference

![image](https://user-images.githubusercontent.com/2443047/36008903-6e5a8314-0cfe-11e8-9062-5a4acfb02742.png)

**Comment by yeoji** [comment]
@appraveen Hmm this is what I got after importing and it's correct.
<img width="1103" alt="screen shot 2018-02-09 at 7 59 10 pm" src="https://user-images.githubusercontent.com/8002179/36019573-d538a22a-0dd3-11e8-8db3-a88e0b08fc52.png">

Is your plugin updated to the latest version? Does the import layout table change at all after importing?

**Comment by e-tyryshkin** [comment]
@yeoji Same issue for 2017.3.4. Doesn't import counts to use "*" as well. Maybe something had changed in the API since 2017.2?

**Comment by yeoji** [comment]
@appraveen @e-tyryshkin You're right, I just upgraded to 2017.3.4 to test this out and it doesn't import properly! Hmm since the plugin is developed with 2016.1, I'm not too sure how to go about figuring out what has changed/fixing this. I'll try and see if I can find anything out though.

**Comment by jshiell** [comment]
This is a standard problem, I fear - sometime new versions can do odd things. But people tend to get a bit miffed if we follow the release, as a lot of large orgs don't or can't upgrade immediately.

I'd try switching the version referenced in `build.gradle` to match the current release and see if anything shows up. Even if we can't release with it built against that SDK, perhaps the results will be illuminating.

**Comment by appraveen** [comment]
This should not be a problem. Looking at the Intellij Version 2017.2.5 to 2017.3.4, API seem to be broken between their `minor` version upgrade. I don't know have any knowledge on their API. Perhaps, you can create an issue with Intellij to check why this is broken on 2017.3.4

If something works in 2017.2.5 i would expect it to work on all `2017.x.y` versions where `x>=2 (y>=5 when x==2, otherwise y can be '*')`

**Comment by mmoayyed** [comment]
I am using the plugin version: `5.20.0` in IDEA 2018.2:

```
IntelliJ IDEA 2018.2 EAP (Ultimate Edition)
Build #IU-182.3684.2, built on July 10, 2018
JRE: 1.8.0_152-release-1248-b8 x86_64
JVM: OpenJDK 64-Bit Server VM by JetBrains s.r.o
macOS 10.13.5
```

Here is an a snippet that demonstrates a CustomImportOrder check:

```xml
<module name="CustomImportOrder">
            <property name="severity" value="error"/>
            <property name="customImportOrderRules"
                      value="SAME_PACKAGE(3)###THIRD_PARTY_PACKAGE###SPECIAL_IMPORTS###STANDARD_JAVA_PACKAGE###STATIC"/>
            <property name="specialImportsRegExp" value="^javax\."/>
            <property name="standardPackageRegExp" value="^java\."/>
            <property name="sortImportsInGroupAlphabetically" value="true"/>
            <property name="separateLineBetweenGroups" value="false"/>
</module>
```

Full example [available here](https://github.com/apereo/cas/blob/7bb9942f9dc4309667d005ff0ebaa1ce6f48b278/style/checkstyle-new-rules.xml)

Importing the configuration from `Preferences / Code Style / Java` presents:

![image](https://user-images.githubusercontent.com/1205228/42677329-ad14b172-8690-11e8-8a84-5fb0029aee09.png)

Then, I changed the rule to the following, taking out `severity`:

```xml
<module name="CustomImportOrder">
            <property name="customImportOrderRules"
                      value="SAME_PACKAGE(3)###THIRD_PARTY_PACKAGE###SPECIAL_IMPORTS###STANDARD_JAVA_PACKAGE###STATIC"/>
            <property name="specialImportsRegExp" value="^javax\."/>
            <property name="standardPackageRegExp" value="^java\."/>
            <property name="sortImportsInGroupAlphabetically" value="true"/>
            <property name="separateLineBetweenGroups" value="false"/>
</module>
```

Second error presents itself complaining about missing enum constant for `SAME_PACKAGE(3)`. So I switched the rule to:

```xml
<module name="CustomImportOrder">
            <property name="customImportOrderRules"
                      value="SAME_PACKAGE###THIRD_PARTY_PACKAGE###SPECIAL_IMPORTS###STANDARD_JAVA_PACKAGE###STATIC"/>
            <property name="specialImportsRegExp" value="^javax\."/>
            <property name="standardPackageRegExp" value="^java\."/>
            <property name="sortImportsInGroupAlphabetically" value="true"/>
            <property name="separateLineBetweenGroups" value="false"/>
</module>
```

In summary and to confirm:

1. Seems like `severity` is not supported by the plugin?
2. Seems like `SAME_PACKAGE(n)` is not supported by the plugin?

PS Happy to open up a separate issue, if it helps further.

**Comment by jshiell** [comment]
@mmoayyed thanks for the report. What version of Checkstyle are you using with this config? At first glance that exception seems to be coming from Checkstyle, not the plugin. In general the plugin just concerns itself with munging the file paths - most of the work, including the majority of parsing, is still done by the underlying Checkstyle instance.

**Comment by mmoayyed** [comment]
Apologies. I should have mentioned that I use Checkstyle `8.11`. (Corrected the original post).

It's worth noting that with the original `CustomImportOrder` rule (with severity and such defined), the plugin is able to perfectly recognize the Checkstyle rules and examine the code, reporting back faulty import orders. That bit works just fine. The issue seems to only present itself when the rules are about to be *imported* into IDEA as a Code Style profile.

**Comment by jshiell** [comment]
@mmoayyed thanks - and another thank you for your excellent bug report, it's much appreciated!

The good news: I can reproduce it. The file works fine for addition & scanning, but fails when imported as a code style. The bad news: the code style work was a contribution, so it's a bit of a mystery. But I'll see what I can dig up 😄

**Comment by jshiell** [comment]
I've added some fixes - `severity` is now supported (and ignored), and `SAME_PACKAGE(n)` should now work (it wasn't checking for the `(n)` previously). Released as 5.21.0.

**Comment by mmoayyed** [comment]
Outstanding. Thank you very much for your time. I'll give the new release a try and will report back.

**Comment by mmoayyed** [comment]
I should report back the latest version of the plugin that is 5.21.0 resolves this issue. Once again, thank you very much for your help. I hope to get better at this gradually and start contributing more :)

**Comment by chriswhite199** [comment]
I'm seeing a similar issue with 5.22.1:

![screen shot 2018-09-21 at 8 29 31 pm](https://user-images.githubusercontent.com/1356204/45912863-4b16af00-bddd-11e8-86a8-5de1ed4ccd6f.png)

    <module name="CustomImportOrder">
        <property name="sortImportsInGroupAlphabetically" value="true"/>
        <property name="separateLineBetweenGroups" value="false"/>
        <property name="customImportOrderRules" value="STATIC###THIRD_PARTY_PACKAGE"/>
    </module>

So the checkstyle.xml is saying static first, but the Import Layout section of the Code Style after importing (with no errors reported) is still showing the default IntelliJ order (statics last and separate java & javax sections).

The Checkstyle plugin is configured with the right version (8.12)

**Comment by romandvoskin-circle** [comment]
This is still an 5.54.0. Here's a snippet of the imports section I was using 
```
<module name="CustomImportOrder">
            <property name="customImportOrderRules"
                      value="THIRD_PARTY_PACKAGE###SPECIAL_IMPORTS###STANDARD_JAVA_PACKAGE###STATIC"/>
            <property name="specialImportsRegExp" value="^javax\."/>
            <property name="standardPackageRegExp" value="^java\."/>
            <property name="sortImportsInGroupAlphabetically" value="true"/>
            <property name="separateLineBetweenGroups" value="false"/>
</module>
```

I found a workaround:

1. Import the checkstyle settings into Code Style | Java | Scheme IDE | Gear | Import Scheme
1. Manually set the `imports` tab | Import Layout until the CheckStyle passes
1. Gear | Export | Editorconfig file  
This exports the resulting settings as `.editorconfig` (to help share with your team)

**Comment by Haarolean** [comment]
This is still an issue in 2021.

> This is still an issue in 5.54.0. Here's a snippet of the imports section I was using
> 
> ```
> <module name="CustomImportOrder">
>             <property name="customImportOrderRules"
>                       value="THIRD_PARTY_PACKAGE###SPECIAL_IMPORTS###STANDARD_JAVA_PACKAGE###STATIC"/>
>             <property name="specialImportsRegExp" value="^javax\."/>
>             <property name="standardPackageRegExp" value="^java\."/>
>             <property name="sortImportsInGroupAlphabetically" value="true"/>
>             <property name="separateLineBetweenGroups" value="false"/>
> </module>
> ```
> 
> I found a workaround:
> 
> 1. Import the checkstyle settings into Code Style | Java | Scheme IDE | Gear | Import Scheme
> 2. Manually set the `imports` tab | Import Layout until the CheckStyle passes
> 3. Gear | Export | Editorconfig file
>    This exports the resulting settings as `.editorconfig` (to help share with your team)

The huge downside is that once you edit the checkstyle again you have to edit the order again.

**Comment by jshiell** [comment]
Added some support for the alpha sort.

**Comment by jshiell** [comment]
Warning added in 26.16.0.
